engine/execution/rpc/engine.go

```diff
@@ -270,18 +270,43 @@ func (h *handler) GetEventsForBlockIDs(
 			return nil, status.Errorf(codes.Internal, "state commitment for block ID %s could not be retrieved", bID)
 		}
 
-		// lookup events
-		blockEvents, err := h.events.ByBlockIDEventType(bID, flow.EventType(eType))
+		// lookup all events for the block
+		blockAllEvents, err := h.events.ByBlockID(bID)
 		if err != nil {
 			return nil, status.Errorf(codes.Internal, "failed to get events for block: %v", err)
 		}
 
+		// filter events by type
+		eventType := flow.EventType(eType)
+		blockEvents := make([]flow.Event, 0, len(blockAllEvents))
+		for _, event := range blockAllEvents {
+			if event.Type == eventType {
+				blockEvents = append(blockEvents, event)
+			}
+		}
+
 		result, err := h.eventResult(bID, blockEvents)
 		if err != nil {
 			return nil, err
 		}
 		results[i] = result
 
+		// additional check that when there is no event in the block, double check if the execution
+		// result has no events as well, otherwise return an error.
+		// we check the execution result has no event by checking if each chunk's EventCollection is
+		// the default hash for empty event collection.
+		if len(blockEvents) == 0 {
```

Comment:
This is checking that `blockEvents` is empty. But it seems like `blockEvents` is the type-filtered list of events, not the list of all events in the block. So we could have a non-empty list of events for the block, but an empty `blockEvents` list, if the queried event type doesn't happen to appear in the block.

I think we should use `blockAllEvents` in this conditional instead. Then, we can also move this conditional block to immediately after `blockAllEvents` is defined on line 278 (fail faster).

Reply:
Good catch!